### PR TITLE
TSFC + Vectorizer

### PR DESCRIPTION
**Pull request by FabioLuporini**
Make COFFEE capable of applying padding and data alignment to TSFC kernels.

This transformation is compasable with all other transformations. However, it is *not* used by default Firedrake (this is activated at level O3, while Firedrake uses O2). The reason is that despite making most loops easily vectorisable, there are compilers (e.g., GCC ver < 5.0) rejecting auto-vectorisation when loops are too small or due to other non-sense reasons. Padding increases the loop trip counts, so it's (performance-wise) safer to switch it on only when sure that loops are gonna get vectorised.

Note: as I'll document properly and some of you already know, there is a non-trivial interplay between the zero-elimination transformation and padding, because of the introduction of offsets in the kernel (e.g., A[i+3]). This, in some cases, makes data alignment not possible. One should try both O2 and O3 when the performance of a specific kernel is really critical.

Ah, the PR also adds a mechanism to pick the proper citations depending on the set of optimisations applied (didn't bother creating a separate PR). As such, this PR has the (trivial) dependency firedrake #796

### CODE REVIEW

File: coffee/citations.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        nozeros = params.get('dead_ops_elimination', False)
+        align_pad = params.get('align_pad', False)
+        vectorize = params.get('vectorize', False)
+        split = params.get('split', False)
+
+        if rewrite == 1 or align_pad or vectorize or split:
+            Citations().register('Luporini2015')
+        if rewrite > 1 or nozeros:
+            Citations().register('Luporini2016')
+    except:
```
Comment (wence-) [inline coffee/citations.py]: You should probably only catch `ImportError`.

Comment (FabioLuporini) [inline coffee/citations.py]: right, fixed

File: coffee/citations.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+# This file is part of COFFEE
+#
+# COFFEE is Copyright (c) 2014, Imperial College London.
```
Comment (miklos1) [inline coffee/citations.py]: Wrong year again.

Comment (FabioLuporini) [inline coffee/citations.py]: Damn...fixed

File: coffee/plan.py [hunk 3]
```diff
@@ -176,6 +178,8 @@ def _generate_cpu_code(self, kernel, **kwargs):
         else:
             params = opts
 
+        update_citations(params)
```
Comment (wence-) [inline coffee/plan.py]: The only bad thing about this approach is that if we hit the firedrake kernel cache, coffee is never called, and so update_citations(params) is never called.

Comment (FabioLuporini) [inline coffee/plan.py]: Mm that's true. I could mimic this logic in Firedrake, but Firedrake only "sees" O2 and O3...and that may become a problem as well....?

OK, in Firedrake I could have (I guess in parameters.py):
O1 -> pick Y
O2 -> pick X
O3 -> pick Y 
if something different than O{1,2,3}, then the user is probably in the business of performance optimisation, so we may (i) just rely on the fact that he will find out what the proper citations are or (ii) issue a warning that recalls to clean the cache to get the citations (clean cache => COFFEE will add the citations through the very same mechanism we have now with this PR)

Other options ? Maybe there's a simpler/better way, but this is what I had on top of my head

Comment (FabioLuporini) [inline coffee/plan.py]: Any idea about this ?

**Comment by FabioLuporini** [comment]
This is now rebased on master

**Comment by FabioLuporini** [comment]
Now depends on PR 817 in Firedrake and PR 492 in PyOP2

**Comment by FabioLuporini** [comment]
This also fixes #8

File: coffee/system.py [hunk 4]
```diff
@@ -85,16 +123,51 @@ def coffee_init(**kwargs):
...
 
     # Allow visits of ASTs that become huge due to transformation. The constant
     # /4000/ was empirically found to be an acceptable value
     sys.setrecursionlimit(4000)
 
 
+def coffee_reconfigure(**kwargs):
+    """Reconfigure the internal state of COFFEE."""
+
+    global options
```
Comment (miklos1) [inline coffee/system.py]: Why do you need this?

Comment (FabioLuporini) [inline coffee/system.py]: My bad, removed

**Comment by FabioLuporini** [comment]
Prod

**Pull request merged**